theano/tensor/nnet/nnet.py

```diff
@@ -746,17 +877,34 @@ def local_logsoftmax(node):
 
     Note: only forward pass is affected
     """
+    list_classes = ((Subtensor, AdvancedSubtensor1, AdvancedSubtensor))
     if (isinstance(node.op, tensor.Elemwise) and
             isinstance(node.op.scalar_op, scalar.basic.Log) and
-            len(node.inputs) == 1 and
-            node.inputs[0].owner is not None and
-            isinstance(node.inputs[0].owner.op, Softmax)):
-        inVars = node.inputs[0].owner.inputs[0]
-        new_op = LogSoftmax()
-        ret = new_op(inVars)
-        ret .tag.values_eq_approx = values_eq_approx_remove_inf
-        copy_stack_trace([node.inputs[0], node.outputs[0]], ret)
-        return [ret]
+            node.inputs[0].owner is not None):
+
+            if isinstance(node.inputs[0].owner.op, Softmax):
+                softmax_input = node.inputs[0].owner.inputs[0]
+                axis = node.inputs[0].owner.op.axis
+                ret = LogSoftmax(axis)(softmax_input)
+                ret.tag.values_eq_approx = values_eq_approx_remove_inf
+                copy_stack_trace([node.inputs[0], node.outputs[0]], ret)
+                return [ret]
+
+            elif isinstance(node.inputs[0].owner.op, list_classes):
+                subtensor_op = node.inputs[0].owner.op
+                subtensor_input = node.inputs[0].owner.inputs[0]
+                subtensor_inputs = node.inputs[0].owner.inputs[1:]
+                if subtensor_input.owner is not None and isinstance(subtensor_input.owner.op, Softmax):
+                    softmax_op = subtensor_input.owner.op
+                    softmax_input = subtensor_input.owner.inputs[0]
+                    # If we have a matrix, we still use the
+                    # CrossentropySoftmaxArgmax1HotWithBias
+                    if softmax_input.type.ndim != 2:
```

Comment:
Yeah, if I don't specify this, the Cross Entropy optimization isn't applied. I will investigate.